pkg/api/latest/latest.go

```diff
@@ -18,125 +18,101 @@ package latest
 
 import (
 	"fmt"
-	"sort"
-	"strings"
 
+	"k8s.io/kubernetes/pkg/api"
 	"k8s.io/kubernetes/pkg/api/meta"
 	"k8s.io/kubernetes/pkg/api/registered"
 	"k8s.io/kubernetes/pkg/api/unversioned"
+	"k8s.io/kubernetes/pkg/api/v1"
 	"k8s.io/kubernetes/pkg/runtime"
+	"k8s.io/kubernetes/pkg/util/sets"
 )
 
-var (
-	allGroups = GroupMetaMap{}
-	// Group is a shortcut to allGroups.Group.
-	Group = allGroups.Group
-	// RegisterGroup is a shortcut to allGroups.RegisterGroup.
-	RegisterGroup = allGroups.RegisterGroup
-	// GroupOrDie is a shortcut to allGroups.GroupOrDie.
-	GroupOrDie = allGroups.GroupOrDie
-	// AllPreferredGroupVersions returns the preferred versions of all
-	// registered groups in the form of "group1/version1,group2/version2,..."
-	AllPreferredGroupVersions = allGroups.AllPreferredGroupVersions
-	// IsRegistered is a shortcut to allGroups.IsRegistered.
-	IsRegistered = allGroups.IsRegistered
-)
+// ExternalVersions lists all known external versions for this group from most preferred to least preferred
+var ExternalVersions = []unversioned.GroupVersion{v1.SchemeGroupVersion}
```

Comment:
ok... looks like what is actually allowed is the intersection of this list's initial value, and what is in registered.supportedVersions (or what is in `$KUBE_API_VERSIONS`, if specified)... which is computed in init(), ordered by whatever is specified in the registered package, and reassigned here.

It is pretty misleading to have an initial list and priority order that gets modified and reshuffled in `init()`. I'd rather see something like:

```
// ExternalVersions lists all known external versions for this group from most preferred to least preferred
var ExternalVersions []unversioned.GroupVersion
// Codec is the Codec for the most preferred ExternalVersion
var Codec runtime.Codec
...
func init() {
  validVersions := map[unversioned.GroupVersion]bool{v1.SchemeGroupVersion: true}
  for _, registeredVersion := range registered.GroupVersionsForGroup(api.SchemeGroupVersion.Group) {
    if !validVersions[registeredVersion] {
      panic(...)
    }
    ExternalVersions = append(ExternalVersions, registeredVersion)
  }

  if len(ExternalVersions) > 0 {
    Codec = runtime.CodecFor(api.Scheme, ExternalVersions[0].String())
    RESTMapper = newRESTMapper()
  }
}
```